### ISSUE TITLE
How to help with pydm

### ISSUE THREAD

**Issue by genderev** [iree-org/iree#7539]
Hi! I would like to help develop [pydm](https://github.com/google/iree/tree/main/llvm-external-projects/iree-dialects/python/iree/compiler/dialects/iree_pydm) but am not sure how to. Where could I start?

I searched for `pydm` inside the [docs](https://google.github.io/iree) and did not find results. I saw this in the UI for the docs after searching for `pydm`: `No matching documents`.

**Comment by stellaraccident** [comment]
As noted on discord:
---
I've been pushing code to here to get bootstrapped with basic features. Will be stopping in the next day or so and working on tests and landing: https://github.com/google/iree/pull/7301
I've been pushing e2e tests to this branch: https://github.com/google/iree-samples/tree/laurenzo_pydm/pydm/simple
I haven't figured out a good place to put them yet since they span some project components.
These things are always hard to get started because you need to connect a certain minimum amount of things before you have much of anything. I'm trying to get to that minimal point.
---

It's not documented because this is highly experimental. I have high hopes for it, but that is what they are right now.

Once I land the above patch (probably next week - I need to write a lot of tests), it may be possible to carve out some areas to work on, but this is all very young with quite a few design decisions either just in my notebook or needing to be worked out.

What level of effort are you willing to put in? And what parts do you think would be good to contribute to?

**Comment by genderev** [comment]
When you say you've been pushing code here to get bootstrapped with basic features, what are those basic features?

What is the purpose of the end to end (I assume that is what e2e stands for) tests?

I'm willing to spend 30-60 minutes working on this each day for the next two months. I'm not sure what to contribute to, because I'm not sure what the the state of the project is.

**Comment by stellaraccident** [comment]
Continuing from discord (deduping conversationg):

The near term goal is end-to-end compilation and execution of simple but semantically meaningful Python programs. This is a bit squishy, but roughly I'm moving the AST-typeinference-runtimelibrary-lowering snowball down the hill piece by piece by focusing on types and concepts that are at the foundation of the language or are tricky and need precise modeling early: functions, numerics, lists, iterators, control flow, variables, assignment, etc. The desire is to have a framework which will compile really nice code in the presence of types and not excessive use of dynamic features but will degrade to full featured dynamic-mode Python.

This is primarily different from Cython, for example, because a) It does not presume the presence of the CPython runtime library, b) it is build as a toolkit in MLIR, and c) we are focused on extracting parts of the program for ahead of time compilation and distribution as part of other systems. We've got a number of things that we want to use it for which would benefit from a domain specific Python compiler toolkit.

The PYDM dialect is *a* modeling of the Python AST in that it is designed to be a trivial transformation from the AST while composing well with what comes below (type inference, lower level types, lowerings, etc). The `importer.py` bridges the gap between Python AST and PYDM dialect. In my experience, this layer has to be codesigned correctly with the needs of the rest of the system and not just be a completely generic modeling of the AST. This influences everything about the dialect design from the types (i.e. boxing is central to PYDM but does not exist at the AST level), exception handling methodology (i.e. the way we model exception_result makes it unnecessary to have non-local jumps, which is a desirable feature further down), variable management (how you model free and cell variables is *really* tricky if you want it to compose), and the specific ops being formulated to work at the right granularity and interop with type inference.

When the above referenced PR lands, PYDM will be able to run simple numeric/loopy/list-y programs like fibonacci generators. My plan is to grow that from there, proceeding down the axis of core concepts that need to be designed early (next up is probably iterator and for..in modeling). When possible, I leave off things that are pure elaboration (i.e. supporting every type of arithmetic over every type in the numeric hierarchy or every method of list) until a later phase, because at this phase, design decisions can still shift, causing a lot of things to need to be invalidated. The big items that I have sketched in my notebook but am still working through are the correct modeling for first class functions, non local state in general, and function calls. These are quite tricky and the cost of getting them wrong is high.

I am spending about 1/3 of my time on this project, and I expect to be able to get most of the big design decision-impacting features nailed down this year. This is a hard time to scale a project to new people unless if they are in sync technically and commitment level. While I'd love to accept the help, it is a hard time to shard the work into easy, incremental pieces unless if the person has done this kind of dynamic language implementation before.

I don't mean to pry, but it sounds like you have an education/curriculum component to what you are wanting to do? If that is the case, I suspect you are looking for a concrete thing you can show in the next couple of months? There may be such a scoped unit here (i.e. with what is there now, you have everything you need to compile fibonacci programs to web-assembly, for example) but I'm focused on a bit longer/holistic time horizon and need to stay so. If you wanted to take some of this and use it/adapt it for simple things in the near term, that might be doable.

**Comment by genderev** [comment]
> We've got a number of things that we want to use it for which would benefit from a domain specific Python compiler toolkit. 

What are those things?

> I suspect you are looking for a concrete thing you can show in the next couple of months?

That would be helpful. I would like to actually contribute instead of just compiling an existing program to WebAssembly.

> This is a hard time to scale a project to new people unless if they are in sync technically and commitment level. 

I think I am more committed than I am technically qualified. Do you recommend anything I can read to be more helpful? I thought [this](https://mlir.llvm.org/docs/Tutorials/Toy/) was useful but I think I need to learn more.

**Comment by jpienaar** [comment]
PyDM has been deprioritized/deleted, so closing for now as no active work there.
